src/move_and_slide.rs

```diff
@@ -99,3 +100,82 @@ pub fn move_and_slide(
         }
     }
 }
+
+fn similar_plane(a: Vec3, b: Vec3) -> bool {
+    a.dot(b) > SIMILARITY_THRESHOLD
+}
+
+fn solve_collision_planes(
+    velocity: Vec3,
+    hits: &[Vec3],
+    original_velocity_direction: Vec3,
+) -> Vec3 {
+    // Early out if we have no velocity or no hits
+    if velocity.length_squared() <= 0.0 || original_velocity_direction.length_squared() <= 0.0 {
+        return Vec3::ZERO;
+    }
+
+    if hits.is_empty() {
+        return velocity;
+    }
+
+    // Do our initial rejection to calculate the sliding velocity.
+    let first_hit_normal = hits[hits.len() - 1];
+    let velocity_direction = velocity.normalize_or_zero();
+    if velocity_direction.dot(first_hit_normal) >= 0.0 {
+        return velocity;
+    }
+    let initial_velocity = velocity.reject_from(first_hit_normal);
+
+    // Join the original velocity direction as an additional constraining plane
+    let original_velocity_normal = original_velocity_direction.normalize_or_zero();
+    let all_hits: Vec<Vec3> = std::iter::once(original_velocity_normal)
+        .chain(hits.iter().cloned())
+        .collect();
+
+    // We should filter out any normals that are similar to the existing constraints
+    let mut filtered_hits = all_hits.iter().filter(|&n| {
+        !similar_plane(first_hit_normal, *n) && !similar_plane(original_velocity_normal, *n)
+    });
+
+    filtered_hits.try_fold(initial_velocity, |vel, second_hit_normal| {
+        let vel = vel.reject_from(*second_hit_normal);
+        let vel_dir = vel.normalize_or_zero();
+
+        // If the velocity is already parallel to the first hit normal, we can return it directly
+        if similar_plane(vel_dir, first_hit_normal) {
+            // If the velocity is small enough we can just assume we have no reason to move
+            if vel.length_squared() <= f32::EPSILON {
+                Err(vel)
+                // Otherwise we need to keep working.
+            } else {
+                Ok(vel)
```

Comment:
```suggestion
                // Otherwise we need to keep working.
            } else {
                Ok(vel)
```
```suggestion
            } else {
                // Otherwise we need to keep working.
                Ok(vel)
```